add-database

7Jenko opened the pull request:
Создание базы данных.
Доработка бизнес-логики.

7Jenko committed:
add-database
Создание базы данных.
Доработка бизнес-логики.

7Jenko committed:
add-database
Создание базы данных.
Доработка бизнес-логики.

7Jenko committed:
add-database
Создание базы данных.
Доработка бизнес-логики.

7Jenko committed:
add-database
Создание базы данных.
Доработка бизнес-логики.

7Jenko committed:
Revert "add-database"

This reverts commit b5dd59efe2318ab1e483fce037a6b111a37f58fc.

7Jenko committed:
add-database
Создание базы данных.
Доработка бизнес-логики.

7Jenko committed:
add-database
Создание базы данных.
Доработка бизнес-логики.

7Jenko committed:
add-database
Создание базы данных.
Доработка бизнес-логики.

7Jenko commented:
Обращение к тому, кто будет проверять.
Уважаемый ревьюер. Помоги, пожалуйста, наведи на верную мысль, не увеличивается количество лайков. Уже третий день бьюсь, не понимаю.

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -48,4 +47,24 @@ public Film(int id, String name, String description, LocalDate releaseDate, Long
         return this.releaseDate.isAfter(MIN_DATE_RELEASE);
     }
+
+    @NotNull(message = "Жанры не могут быть пустыми.")
```
Чтобы провалидировать, что коллекция не пустая, следует указать `@NotEmpty`

7Jenko replied:
@slavrikk Привет, при смене на @NotEmpty, сыпятся 16 тестов. Могу оставить @NotNull, пожалуйста?
![16 failed tests](https://github.com/user-attachments/assets/b7dd2160-5213-4867-b60d-f0e3441b3870)

slavrikk replied:
хорошо

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -2,16 +2,16 @@
-import java.util.stream.Collectors;
 
-@Slf4j
+@Component
```
`@Component` лишний, так как уже есть `@Service`

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,238 @@
+
+    private Set<Genre> getGenres(int filmId) {
+        Comparator<Genre> compId = Comparator.comparing(Genre::getId);
+        Set<Genre> genres = new TreeSet<>(compId);
```
Кажется `TreeSet` и `Comparator` лишний, так как все жанры будут отсортированы в sql запросе, массива будет достаточно

7Jenko replied:
та же история с TreeSet и Comparator, падают все тесты, касающиеся жанров.
![43 failed tests](https://github.com/user-attachments/assets/bb912579-f56d-4219-824e-68841befa24d)

slavrikk replied:
хорошо

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,238 @@
+                .build();
+    }
+
+    private Film filmMap(SqlRowSet srs) {
```
Кажется методы `makeFilm()` и `filmMap()` делают одно и тоже только из разных источников, можно попробовать унифицировать

slavrikk commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserStorage.java:
```diff
@@ -1,10 +1,11 @@
 import java.util.List;
 
-public interface UserStorage {
+public interface UserStorage extends FriendsStorage {
```
`FriendsStorage` и `UserStorage` имеют разные источники данных, лучше их не наследовать

slavrikk commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,55 @@
+);
+
+ALTER TABLE films ALTER COLUMN rating_id SET DEFAULT 1;
+ALTER TABLE films ALTER COLUMN rating_id DROP NOT NULL;
```
Можно сразу указать эти ограничения при создании таблицы

slavrikk reviewed (commented):
Также не проходят тесты postman, посмотри пожалуйста

7Jenko committed:
add-database
Указать ограничения при создании таблицы.

7Jenko committed:
add-database
Унифицировать методы makeFilm() и filmMap().

7Jenko committed:
add-database
Убрать @Component в UserService.

slavrikk reviewed (commented):
Осталось поправить один тест

7Jenko commented:
> Осталось поправить один тест

@slavrikk подскажи, пожалуйста, в чем может быть проблема? я вижу по логам, что у меня не увеличивается количество лайков на фильмах. Уже пробовал отделить логику лайков в отдельные классы, но еще хуже становилось. Буду благодарен, если наведешь на верную мысль.

7Jenko committed:
add-database
Разделить логику like от Film.

7Jenko committed:
add-database
Разделить логику like от Film.

7Jenko committed:
#  failure                                                       detail

 1.  AssertionError                                                Test film 'genres' field
                                                                   "genres" field must be %{requestGenres}: expected [ 5, 3 ] to deeply equal [ 3, 5 ]
                                                                   at assertion:8 in test-script
                                                                   inside "films / Film create several genres"

7Jenko committed:
add-database
Tests passed.

7Jenko committed:
add-database
Tests passed.
